### PR TITLE
Add the ability to output mesh info after mesh generator execution

### LINKED ISSUE #16783: Add ability to output detailed information about a MeshGenerator
## Reason
It would be useful to know detailed information about meshes that are generated by a MeshGenerator. Information like the bounding box, sidesets, nodesets, subdomains, etc.

## Design
Add a `show_info` parameter (name tbd) to `MeshGenerator` that prints this information.

## Impact
Will help with debugging mesh generation using MeshGenerators.

### LINKED ISSUE #17045: PatternedMeshGenerator fails with distributed input meshes
## Reason
```
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
  []
  [pmg]
    type = PatternedMeshGenerator
    inputs = 'gmg'
    pattern = '0'
  []
[]
```

will fail with a seg fault if ran with multiple procs and distributed mesh, because it casts inputs to ReplicatedMesh and doesn't check if the cast failed.

## Design
Add a more useful error if the input meshes aren't replicated.

## Impact
Better errors.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #16783 and closes #17045.

Here's an example with the following:

```
[Mesh]
  [gmg_quad]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 4
    ny = 4
    show_info = true
  []
  [gmg_quad_block1]
    type = ParsedSubdomainMeshGenerator
    input = gmg_quad
    combinatorial_geometry = 'x > 0.5'
    block_name = 'dummy'
    block_id = 1
    show_info = true
  []
[]
```

Output:

```
GeneratedMeshGenerator 'gmg_quad':  Mesh Information:
GeneratedMeshGenerator 'gmg_quad':   elem_dimensions()={2}
GeneratedMeshGenerator 'gmg_quad':   spatial_dimension()=2
GeneratedMeshGenerator 'gmg_quad':   n_nodes()=25
GeneratedMeshGenerator 'gmg_quad':     n_local_nodes()=25
GeneratedMeshGenerator 'gmg_quad':   n_elem()=16
GeneratedMeshGenerator 'gmg_quad':     n_local_elem()=16
GeneratedMeshGenerator 'gmg_quad':     n_active_elem()=16
GeneratedMeshGenerator 'gmg_quad':   n_subdomains()=1
GeneratedMeshGenerator 'gmg_quad':   n_partitions()=1
GeneratedMeshGenerator 'gmg_quad':   n_processors()=1
GeneratedMeshGenerator 'gmg_quad':   n_threads()=1
GeneratedMeshGenerator 'gmg_quad':   processor_id()=0
GeneratedMeshGenerator 'gmg_quad':   is_prepared()=true
GeneratedMeshGenerator 'gmg_quad':   is_replicated()=true
GeneratedMeshGenerator 'gmg_quad':
GeneratedMeshGenerator 'gmg_quad':  Mesh Bounding Box:
GeneratedMeshGenerator 'gmg_quad':   Minimum: (x,y,z)=(       0,        0,        0)
GeneratedMeshGenerator 'gmg_quad':   Maximum: (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':   Delta:   (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':
GeneratedMeshGenerator 'gmg_quad':  Mesh Element Type(s):
GeneratedMeshGenerator 'gmg_quad':   QUAD4
GeneratedMeshGenerator 'gmg_quad':
GeneratedMeshGenerator 'gmg_quad':  Mesh Nodesets:
GeneratedMeshGenerator 'gmg_quad':   Nodeset 0 (bottom), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':   Nodeset 1 (right), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':   Nodeset 2 (top), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':   Nodeset 3 (left), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':
GeneratedMeshGenerator 'gmg_quad':  Mesh Sidesets:
GeneratedMeshGenerator 'gmg_quad':   Sideset 0 (bottom), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Side volume: 1
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':   Sideset 1 (right), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Side volume: 1
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':   Sideset 2 (top), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Side volume: 1
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       1,        0,        0)
GeneratedMeshGenerator 'gmg_quad':   Sideset 3 (left), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
GeneratedMeshGenerator 'gmg_quad':    Side volume: 1
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       0,        1,        0)
GeneratedMeshGenerator 'gmg_quad':
GeneratedMeshGenerator 'gmg_quad':  Mesh Edgesets:
GeneratedMeshGenerator 'gmg_quad':   None
GeneratedMeshGenerator 'gmg_quad':
GeneratedMeshGenerator 'gmg_quad':  Mesh Subdomains:
GeneratedMeshGenerator 'gmg_quad':   Subdomain 0: 16 elems (QUAD4, 16 active), 25 active nodes
GeneratedMeshGenerator 'gmg_quad':    Volume: 1
GeneratedMeshGenerator 'gmg_quad':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':    Bounding box delta: (x,y,z)=(       1,        1,        0)
GeneratedMeshGenerator 'gmg_quad':   Global mesh volume = 1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':  Mesh Information:
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   elem_dimensions()={2}
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   spatial_dimension()=2
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   n_nodes()=25
ParsedSubdomainMeshGenerator 'gmg_quad_block1':     n_local_nodes()=25
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   n_elem()=16
ParsedSubdomainMeshGenerator 'gmg_quad_block1':     n_local_elem()=16
ParsedSubdomainMeshGenerator 'gmg_quad_block1':     n_active_elem()=16
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   n_subdomains()=2
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   n_partitions()=1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   n_processors()=1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   n_threads()=1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   processor_id()=0
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   is_prepared()=true
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   is_replicated()=true
ParsedSubdomainMeshGenerator 'gmg_quad_block1':
ParsedSubdomainMeshGenerator 'gmg_quad_block1':  Mesh Bounding Box:
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Minimum: (x,y,z)=(       0,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Maximum: (x,y,z)=(       1,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Delta:   (x,y,z)=(       1,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':
ParsedSubdomainMeshGenerator 'gmg_quad_block1':  Mesh Element Type(s):
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   QUAD4
ParsedSubdomainMeshGenerator 'gmg_quad_block1':
ParsedSubdomainMeshGenerator 'gmg_quad_block1':  Mesh Nodesets:
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Nodeset 0 (bottom), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Nodeset 1 (right), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Nodeset 2 (top), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Nodeset 3 (left), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':
ParsedSubdomainMeshGenerator 'gmg_quad_block1':  Mesh Sidesets:
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Sideset 0 (bottom), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Side volume: 1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Sideset 1 (right), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Side volume: 1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Sideset 2 (top), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Side volume: 1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       1,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Sideset 3 (left), 4 sides (EDGE2), 4 elems (QUAD4), 5 nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Side volume: 1
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(       0,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':
ParsedSubdomainMeshGenerator 'gmg_quad_block1':  Mesh Edgesets:
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   None
ParsedSubdomainMeshGenerator 'gmg_quad_block1':
ParsedSubdomainMeshGenerator 'gmg_quad_block1':  Mesh Subdomains:
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Subdomain 0: 8 elems (QUAD4, 8 active), 15 active nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Volume: 0.5
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(     0.5,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(     0.5,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Subdomain 1 (dummy): 8 elems (QUAD4, 8 active), 15 active nodes
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Volume: 0.5
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box minimum: (x,y,z)=(     0.5,        0,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box maximum: (x,y,z)=(       1,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':    Bounding box delta: (x,y,z)=(     0.5,        1,        0)
ParsedSubdomainMeshGenerator 'gmg_quad_block1':   Global mesh volume = 1
```

### CODE REVIEW

**Comment by loganharbour** [comment]
This will be ready when https://github.com/libMesh/libmesh/pull/2840 gets into libMesh and we have a libMesh update.

File: framework/doc/content/source/meshgenerators/MeshGenerator.md [hunk 1]
```diff
@@ -79,15 +41,102 @@ For instance, take a look at the following input file:
...
+    ny = 2
+    show_info = true
+  []
+```
+
+The above will result in on-screen output similar to the following (ran with two processes):
 
-Let's analyze the different steps. First, we use a `GeneratedMeshGenerator` to create a 3D mesh labeled `gmg` (here it's simply a regular cube). After that, we use this mesh as an input (via `input = gmg`) to create a larger mesh using a `TiledMeshGenerator`.
+```
+GeneratedMeshGenerator 'gmg':  Mesh Information:
```
Comment (aeslaughter) [inline framework/doc/content/source/meshgenerators/MeshGenerator.md:83]: After looking at this, I am not sure I like the prefix. Can you just put the type and name at the beginning of the mesh info dump. Either has a heading or in the information list (or both)

```
\\ MyGenerator::foo \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
Mesh Information:
   Type: MyGenerator
   Name: foo
...
```

**Review by aeslaughter — APPROVED** [review]
If you could answer my format question that would be great. If you have a strong opinion with the current format then I am fine.

File: framework/doc/content/source/meshgenerators/MeshGenerator.md (hunk 1 above)
Comment (loganharbour) [reply framework/doc/content/source/meshgenerators/MeshGenerator.md:83]: I thought about this too, but when you have a ton of MGs that print output it's really nice to be able to search for them easily and see when one ends and one begins by just scrolling and seeing the difference in width on the left.

**Pull request merged**